From 4f331280f65fdddd8095b4b1d98dceb6f9add009 Mon Sep 17 00:00:00 2001
From: Till Schneidereit <till@tillschneidereit.net>
Date: Mon, 26 Jul 2021 11:28:40 +0200
Subject: [PATCH] JSAPI support for resetting the seed for Math.random()

In long-running environments, it's desirable for the embedding to reset the random seed at controlled points in time.

In particular, when running in WASI, a situation can occur where a VM snapshot is taken at build time using [Wizer](https://github.com/bytecodealliance/wizer), which is then loaded and used as the basis for running separate workloads many times over. If `Math.random()` was called during the initialization phase, this would cause all workloads to get the same results for all consecutive calls to `Math.random()`.
---
diff --git a/js/src/builtin/Math.cpp b/js/src/builtin/Math.cpp
index 45d0b27..1eb2bb5 100644
--- a/js/src/builtin/Math.cpp
+++ b/js/src/builtin/Math.cpp
@@ -541,6 +541,12 @@ Realm::getOrCreateRandomNumberGenerator() {
   return randomNumberGenerator_.ref();
 }
 
+void Realm::resetRandomNumberGenerator() {
+  if (randomNumberGenerator_.isSome()) {
+    randomNumberGenerator_.reset();
+  }
+}
+
 double js::math_random_impl(JSContext* cx) {
   return cx->realm()->getOrCreateRandomNumberGenerator().nextDouble();
 }
diff --git a/js/src/jsfriendapi.h b/js/src/jsfriendapi.h
index 8a91563..8db0a3c 100644
--- a/js/src/jsfriendapi.h
+++ b/js/src/jsfriendapi.h
@@ -188,6 +188,14 @@ extern JS_PUBLIC_API void RestartDrainingJobQueue(JSContext* cx);
 
 extern JS_PUBLIC_API void RunJobs(JSContext* cx);
 
+/**
+ * Reset the seed for Math.random().
+ *
+ * Enables embedders to reset the seed at controlled points, e.g. after
+ * resuming execution from an instance snapshot of SpiderMonkey's VM.
+ */
+extern JS_PUBLIC_API void ResetMathRandomSeed(JSContext* cx);
+
 extern JS_PUBLIC_API JS::Zone* GetRealmZone(JS::Realm* realm);
 
 using PreserveWrapperCallback = bool (*)(JSContext*, JS::HandleObject);
diff --git a/js/src/vm/JSContext.cpp b/js/src/vm/JSContext.cpp
index 5156259..3e080d7 100644
--- a/js/src/vm/JSContext.cpp
+++ b/js/src/vm/JSContext.cpp
@@ -851,6 +851,11 @@ bool InternalJobQueue::getHostDefinedData(
   return true;
 }
 
+JS_PUBLIC_API void js::ResetMathRandomSeed(JSContext* cx) {
+  MOZ_ASSERT(cx->realm());
+  cx->realm()->resetRandomNumberGenerator();
+}
+
 void InternalJobQueue::runJobs(JSContext* cx) {
   if (draining_ || interrupted_) {
     return;
diff --git a/js/src/vm/Realm.h b/js/src/vm/Realm.h
index c60b7da..7d47a90 100644
--- a/js/src/vm/Realm.h
+++ b/js/src/vm/Realm.h
@@ -837,6 +837,10 @@ class JS::Realm : public JS::shadow::Realm {
   // Initializes randomNumberGenerator if needed.
   mozilla::non_crypto::XorShift128PlusRNG& getOrCreateRandomNumberGenerator();
 
+  // Resets randomNumberGenerator to null, so it'll be reinitialized on next
+  // use.
+  void resetRandomNumberGenerator();
+
   const mozilla::non_crypto::XorShift128PlusRNG*
   addressOfRandomNumberGenerator() const {
     return randomNumberGenerator_.ptr();
